### PR TITLE
Pilatus setup

### PR DESCRIPTION
**Pull request by gfabbris**
These are a few small tweaks to the Pilatus setup that I expect to make it work. My main uncertainty is whether the handler will work.

I like to have the `default_settings` function can be used to restore the settings.

We should test it at the beamline before merging.

### CODE REVIEW

**Comment by prjemian** [comment]
I believe the `AD_HDF5` handler is already registered.  See: https://github.com/bluesky/databroker/search?q=register_handler

**Comment by prjemian** [comment]
But you are right @gfabbris , it is not clear if the `area_detector_handlers` package is being used: https://github.com/search?q=org%3Abluesky+AreaDetectorHDF5Handler&type=code

**Comment by gfabbris** [comment]
https://pypi.org/project/area-detector-handlers/

File: instrument/devices/pilatus_100k.py [hunk 1]
```diff
@@ -15,47 +14,57 @@
...
-     """Pilatus detector"""
+    """Pilatus detector"""
 
-     image = ADComponent(ImagePlugin, "image1:")
-     hdf1 = ADComponent(
+    _default_configuration_attrs = (
+        'roi1', 'roi2', 'roi3', 'roi4', 'cam'
+    )
+    _default_read_attrs = (
+        'hdf1', 'stats1', 'stats2', 'stats3', 'stats4'
```
Comment (jwkim-anl) [inline instrument/devices/pilatus_100k.py]: add stats5 as well

**Comment by prjemian** [comment]
@danielballan confirmed on Slack:

> @dallan (NSLS-II): Thanks.  You said that, once the area_detector_handlers package is installed, it configures itself to override the matching support provided in databroker.  Right?

**Comment by prjemian** [comment]
This [notebook](https://github.com/BCDA-APS/bluesky_training/blob/main/locate_image_peak.ipynb) uses databroker to read the image from the last run and do some math on it.  You can use this as a starting point to perform processing on your image.

**Comment by gfabbris** [comment]
This is ready for review.

@jwkim-anl: you should take a look at the `.default_configuration` and `.plot_roi` functions to see if that makes sense for you.

@prjemian: we had to use a custom Pilatus handler because the default one did not work. This was a similar issue I had with the Lambda detector, and I'm not entirely sure why this solution worked.

**Comment by prjemian** [comment]
Got this error with the standard `AD_HDF5` handler:

```py
TypeError: __init__() got an unexpected keyword argument 'template'
```

**Comment by prjemian** [comment]
<details>
<summary>full exception trace</summary>

```py
In [69]: run.primary.read()
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:874, in Filler._get_handler_maybe_cached(self, resource)
    873 try:
--> 874     handler = self._handler_cache[key]
    875 except KeyError:

KeyError: ('d8ad5309-629f-44aa-b370-8ab1c7f40703', 'AD_HDF5')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Input In [69], in <cell line: 1>()
----> 1 run.primary.read()

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/databroker/core.py:1371, in BlueskyEventStream.read(self)
   1364 """
   1365 Return data from this Event Stream as an xarray.Dataset.
   1366 
   1367 This loads all of the data into memory. For delayed ("lazy"), chunked
   1368 access to the data, see :meth:`to_dask`.
   1369 """
   1370 # Implemented just so we can put in a docstring
-> 1371 return super().read()

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/databroker/intake_xarray_core/base.py:48, in DataSourceMixin.read(self)
     46 def read(self):
     47     """Return a version of the xarray with all the data in memory"""
---> 48     self._load_metadata()
     49     return self._ds.load()

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/intake/source/base.py:236, in DataSourceBase._load_metadata(self)
    234 """load metadata only if needed"""
    235 if self._schema is None:
--> 236     self._schema = self._get_schema()
    237     self.dtype = self._schema.dtype
    238     self.shape = self._schema.shape

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/databroker/intake_xarray_core/base.py:20, in DataSourceMixin._get_schema(self)
     17 self.urlpath = self._get_cache(self.urlpath)[0]
     19 if self._ds is None:
---> 20     self._open_dataset()
     22     metadata = {
     23         'dims': dict(self._ds.dims),
     24         'data_vars': {k: list(self._ds[k].coords)
     25                       for k in self._ds.data_vars.keys()},
     26         'coords': tuple(self._ds.coords.keys()),
     27     }
     28     if getattr(self, 'on_server', False):

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/databroker/core.py:1350, in BlueskyEventStream._open_dataset(self)
   1336     self._ds = documents_to_xarray_config(
   1337         object_name=self._configuration_for,
   1338         sub_dict=self._sub_dict,
   (...)
   1347         include=self.include,
   1348         exclude=self.exclude)
   1349 else:
-> 1350     self._ds = documents_to_xarray(
   1351         sub_dict=self._sub_dict,
   1352         start_doc=self._run_start_doc,
   1353         stop_doc=self._run_stop_doc,
   1354         descriptor_docs=self._descriptors,
   1355         get_event_pages=self._get_event_pages,
   1356         filler=self.fillers['delayed'],
   1357         get_resource=self._get_resource,
   1358         lookup_resource_for_datum=self._lookup_resource_for_datum,
   1359         get_datum_pages=self._get_datum_pages,
   1360         include=self.include,
   1361         exclude=self.exclude)

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/bluesky_live/conversion.py:87, in documents_to_xarray(start_doc, stop_doc, descriptor_docs, get_event_pages, filler, get_resource, lookup_resource_for_datum, get_datum_pages, include, exclude, sub_dict)
     85     filled_events = []
     86     for event in events:
---> 87         filled_event = _fill(
     88             filler,
     89             event,
     90             lookup_resource_for_datum,
     91             get_resource,
     92             get_datum_pages,
     93         )
     94         filled_events.append(filled_event)
     95 else:

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/bluesky_live/conversion.py:292, in _fill(filler, event, lookup_resource_for_datum, get_resource, get_datum_pages, last_datum_id)
    283 def _fill(
    284     filler,
    285     event,
   (...)
    289     last_datum_id=None,
    290 ):
    291     try:
--> 292         _, filled_event = filler("event", event)
    293         return filled_event
    294     except event_model.UnresolvableForeignKeyError as err:

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:1020, in Filler.__call__(self, name, doc, validate)
   1017 if self._closed:
   1018     raise EventModelRuntimeError(
   1019         "This Filler has been closed and is no longer usable.")
-> 1020 return super().__call__(name, doc, validate)

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:113, in DocumentRouter.__call__(self, name, doc, validate)
     95 def __call__(self, name, doc, validate=False):
     96     """
     97     Process a document.
     98 
   (...)
    111         instance as doc, a copy of doc, or a different dict altogether.
    112     """
--> 113     return self._dispatch(name, doc, validate)

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:121, in DocumentRouter._dispatch(self, name, doc, validate)
    115 def _dispatch(self, name, doc, validate):
    116     """
    117     Dispatch to the method corresponding to the `name`.
    118 
    119     Optionally validate that the result is still a valid document.
    120     """
--> 121     output_doc = getattr(self, name)(doc)
    123     # If 'event' is not defined by the subclass but 'event_page' is, or
    124     # vice versa, use that. And the same for 'datum_page' / 'datum.
    125     if output_doc is NotImplemented:

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:801, in Filler.event(self, doc)
    800 def event(self, doc):
--> 801     filled_doc = self.fill_event(doc, include=self.include,
    802                                  exclude=self.exclude)
    803     return filled_doc

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:947, in Filler.fill_event(self, doc, include, exclude, inplace)
    945 self._current_state.resource = resource
    946 self._current_state.datum = datum_doc
--> 947 handler = self._get_handler_maybe_cached(resource)
    948 error_to_raise = DataNotAccessible(
    949         f"Filler was unable to load the data referenced by "
    950         f"the Datum document {datum_doc} and the Resource "
    951         f"document {resource}.")
    952 payload = _attempt_with_retries(
    953     func=handler,
    954     args=(),
   (...)
    957     error_to_catch=IOError,
    958     error_to_raise=error_to_raise)

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:876, in Filler._get_handler_maybe_cached(self, resource)
    874     handler = self._handler_cache[key]
    875 except KeyError:
--> 876     handler = self.get_handler(resource)
    877     self._handler_cache[key] = handler
    878 return handler

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:861, in Filler.get_handler(self, resource)
    858     msg += (f"Its 'root' field {original_root} was "
    859             f"*not* modified by root_map.")
    860 error_to_raise = EventModelError(msg)
--> 861 handler = _attempt_with_retries(
    862     func=handler_class,
    863     args=(resource_path,),
    864     kwargs=resource['resource_kwargs'],
    865     intervals=[0] + self.retry_intervals,
    866     error_to_catch=IOError,
    867     error_to_raise=error_to_raise)
    868 return handler

File ~/.conda/envs/bluesky_2022_2/lib/python3.9/site-packages/event_model/__init__.py:1041, in _attempt_with_retries(func, args, kwargs, intervals, error_to_catch, error_to_raise)
   1039 ttime.sleep(interval)
   1040 try:
-> 1041     return func(*args, **kwargs)
   1042 except error_to_catch as error_:
   1043     # The file may not be visible on the filesystem yet.
   1044     # Wait and try again. Stash the error in a variable
   1045     # that we can access later if we run out of attempts.
   1046     error = error_

TypeError: __init__() got an unexpected keyword argument 'template'
```

</details>

**Comment by prjemian** [comment]
Probably we should resolve this issue with the standard handler.  When we use a custom `AD_HDF5_Pilatus_6idb` handler, we must make that available in every client that wishes to access the image through tiled/databroker/...

**Comment by prjemian** [comment]
This custom code:

```py
from area_detector_handlers.handlers import AreaDetectorHDF5SingleHandler
from dask.array import from_array
class PilatusHDF5Handler(AreaDetectorHDF5SingleHandler):
    specs = (
        {"AD_HDF5_Pilatus_6idb"} | AreaDetectorHDF5SingleHandler.specs
    )
    def __call__(self, point_number):
        return from_array(super().__call__(point_number))
cat.register_handler("AD_HDF5_Pilatus_6idb", PilatusHDF5Handler, overwrite=True)
```

**Comment by prjemian** [comment]
Similar to an issue in apstools: https://github.com/BCDA-APS/apstools/issues/651

**Comment by gfabbris** [comment]
In setting up this, I based myself in the [`area_detector_handlers.AreaDetectorHDF5SingleHandler`](https://github.com/bluesky/area-detector-handlers/blob/508c9f9ed0ea89cdf85910bddeae1bb710a6bef1/area_detector_handlers/handlers.py#L545), which is specific for the case of 1 image per file. But, in my experience/understanding, there is a bug in this handler, because it outputs a `list` instead of a `dask.array` ([here](https://github.com/bluesky/area-detector-handlers/blob/508c9f9ed0ea89cdf85910bddeae1bb710a6bef1/area_detector_handlers/handlers.py#L536)). That is why I had to explicitly convert the result of the `__call__` method to `dask.array`.

There might be a way to use the `area_detector_handlers.AreaDetectorHDF5Handler` instead, but I never tried since I believe it's designed to be used in "capture" mode. It may work anyway, but we need to change [this](https://github.com/gfabbris/6idb-bluesky-instrument/blob/f2dd8f98944427b74db20d68ed925cbbba5d40a7/instrument/devices/pilatus_100k.py#L23) to use the `FileStoreHDF5IterativeWrite`, and maybe change the `datum_kwargs` and staging process appropriately.

**Pull request closed without merging**